crates/wasi/src/preview2/ctx.rs

```diff
@@ -258,3 +258,53 @@ pub struct WasiCtx {
     pub(crate) stdout: u32,
     pub(crate) stderr: u32,
 }
+
+impl WasiCtx {
+    /// Flush all buffered output and wait for it to reach the destination.
+    ///
+    /// NOTE: This function should be used when [`WasiCtx`] is used in an async embedding
+    /// (i.e. with [`crate::preview2::command::add_to_linker`]). Use its counterpart
+    /// `sync_join_background_tasks` in a synchronous embedding (i.e. with
+    /// [`crate::preview2::command::sync::add_to_linker`].
+    ///
+    /// In order to implement non-blocking streams, we often often need to offload async
+    /// operations to background `tokio::task`s. These tasks are aborted when the resources
+    /// in the `Table` referencing them are dropped. In some cases, this abort may occur before
+    /// buffered output has been flushed. Use this function to wait for all
+    /// written data to reach its destination gracefully, even if wasm didn't
+    /// explicitly wait for this.
+    ///
+    /// In some embeddings, a misbehaving client might cause this graceful exit to await for an
+    /// unbounded amount of time, so we recommend bounding this with a timeout or other mechanism.
+    pub async fn flush_output(&mut self, table: &mut Table) {
+        let keys = table.keys().cloned().collect::<Vec<u32>>();
+        for k in keys {
+            match table.delete_output_stream(k) {
+                Ok(mut ostream) => {
+                    let _ = ostream.flush().await;
+                }
+                _ => {}
+            }
+        }
+    }
+
+    /// Same as [`WasiCtx::flush_output`] except suitable for synchronous
+    /// embeddings.
+    ///
+    /// This will block the current thread up to the `timeout` specified, or
+    /// forever if `None` is specified, until all wasm output has been flushed
+    /// out.
+    pub fn sync_join_background_tasks<'a>(
+        &mut self,
+        table: &'a mut Table,
+        timeout: Option<std::time::Duration>,
+    ) {
+        crate::preview2::in_tokio(async move {
+            if let Some(timeout) = timeout {
+                let _ = tokio::time::timeout(timeout, self.flush_output(table)).await;
+            } else {
+                self.flush_output(table).await
+            }
+        })
+    }
```

Comment:
lets keep this change out please, if possible... even if you need to just sleep(100ms) for now to get stdout/stderr to flush, I want to take the time to get the streams design right rather than merge it in and then back it out immediately.